flink-core/src/main/java/org/apache/flink/configuration/ConfigOptionsDocGenerator.java

```diff
@@ -0,0 +1,279 @@
+	 * @return string containing HTML formatted table
+	 */
+	static String create(Class<?> clazzz) {
+		final List<ConfigOption> configOptions = extractConfigOptions(clazzz);
+
+		return toHTMLTable(configOptions);
+	}
+
+	private static List<ConfigOption> extractConfigOptions(Class<?> clazzz) {
+		try {
+			final List<ConfigOption> configOptions = new ArrayList<>();
+			final Field[] fields = clazzz.getFields();
+			for (Field field : fields) {
+				if (field.getType().equals(ConfigOption.class) && field.getAnnotation(Deprecated.class) == null) {
+					configOptions.add((ConfigOption) field.get(null));
+				}
+			}
+
+			return configOptions;
+		} catch (Exception e) {
+			throw new RuntimeException("Could not retrieve all options.", e);
+		}
+	}
+
+	/**
+	 * Transforms option to table row.
+	 *
+	 * @param option option to transform
+	 * @return row with the option description
+	 */
+	private static String toHTMLString(final ConfigOption<?> option) {
+		return "<tr>" +
+				"<td>" + option.key() + "</td>" +
+				"<td>" + defaultValueToHtml(option.defaultValue()) + "</td>" +
+				"<td>" + option.description() + "</td>" +
+				"</tr>";
+	}
+
+	private static String defaultValueToHtml(Object value) {
+		if (value instanceof String) {
+			return "\"" + value +"\"";
+		}
+
+		return value == null ? "(none)" : value.toString();
+	}
+
+	private static void sortOptions(List<ConfigOption> configOptions) {
+		Collections.sort(configOptions, new Comparator<ConfigOption>() {
+			@Override
+			public int compare(ConfigOption o1, ConfigOption o2) {
+				return o1.key().compareTo(o2.key());
+			}
+		});
+	}
+
+	static List<Tuple2<ConfigGroup, String>> generateTablesForClass(Class<?> optionsClass) {
+		final ConfigGroups configGroups = optionsClass.getAnnotation(ConfigGroups.class);
+		List<Tuple2<ConfigGroup, String>> tables = new ArrayList<>();
+		final List<ConfigOption> allOptions = extractConfigOptions(optionsClass);
+
+		if (configGroups != null) {
+			final Tree tree = new Tree();
+
+			for (ConfigGroup group : configGroups.groups()) {
+				tree.set(group);
+			}
+
+			for (ConfigOption configOption : allOptions) {
+				tree.add(configOption);
+			}
+
+			for (ConfigGroup group : configGroups.groups()) {
+				List<ConfigOption> configOptions = tree.findConfigOptions(group);
+				sortOptions(configOptions);
+				tables.add(Tuple2.of(group, toHTMLTable(configOptions)));
+			}
+			List<ConfigOption> configOptions = tree.getDefaultOptions();
+			sortOptions(configOptions);
+			tables.add(Tuple2.<ConfigGroup, String>of(null, toHTMLTable(configOptions)));
+		} else {
+			sortOptions(allOptions);
+			tables.add(Tuple2.<ConfigGroup, String>of(null, toHTMLTable(allOptions)));
+		}
+		return tables;
+	}
+
+	/**
+	 * Method used to generated documentation entries containing tables of configuration options with default value and
+	 * description. Each found class matching a pattern *Options.java will results in a separate file with a configuration
```

Comment:
This needs to be updated in regards to `ConfigGroups`, as we can now create multiple files.